socs/agents/devantech_dS378/drivers.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python3
+"""dS378 ethernet relay"""
+import sys
+from enum import IntEnum
+
+from socs.tcp import TCPInterface
+
+# Command byte, length of expected response
+GET_STATUS = 0x30, 8
+SET_RELAY = 0x31, 1
+SET_OUTPUT = 0x32, 1
+GET_RELAYS = 0x33, 5
+GET_INPUTS = 0x34, 2
+GET_ANALOG = 0x35, 14
+GET_COUNTERS = 0x36, 8
+
+# Number of trial for send / receive
+N_TRIAL = 2
+LEN_BUFF = 1024
+
+
+class RelayStatus(IntEnum):
+    """Relay status"""
+    on = 1
+    off = 0
+
+
+class DS378(TCPInterface):
+    """dS378 ethernet relay"""
+
+    def __init__(self, ip, port, timeout=10):
+        super().__init__(ip, port, timeout)
+
+    def __del__(self):
+        self._com.close()
+
+    def _send(self, msg):
+        self.send(msg)
+
+    def _send1(self, msg_byte):
+        self._send(bytearray([msg_byte]))
+
+    def _recv(self):
+        rcv = self.recv(LEN_BUFF)
+        return rcv
+
+    def _recv1(self):
+        return self._recv()[0]
+
+    def _send_recv(self, msg, length):
+        for _ in range(N_TRIAL):
+            self._send(msg)
+            msg_rcv = self._recv()
+            if len(msg_rcv) == length:
+                return msg_rcv
+
+        raise ConnectionError
```

Comment:
This looks like the code [that you mentioned](https://github.com/simonsobs/socs/pull/694#discussion_r1908217838) is handling dropping partial responses when reconnecting. Do you mind just adding a comment here describing what problem this is solving. (The `TCPInterface` class also tries twice before giving up, so without some context this might look a little strange to someone looking at this down the road.)